File: src/main/java/io/realworld/api/UserAndAuthenticationApiImpl.java
package io.realworld.api;

import io.realworld.model.*;
import io.realworld.model.request.CreateUser;
import io.realworld.model.request.Login;
import io.realworld.model.request.UpdateUser;
import io.vertx.core.Future;
import io.vertx.ext.web.RoutingContext;
import io.vertx.ext.web.handler.HttpException;

// Implement this class

public class UserAndAuthenticationApiImpl implements UserAndAuthenticationApi {
  public Future<ApiResponse<Login200Response>> createUser(CreateUser body) {
    return Future.failedFuture(new HttpException(501));
  }

  public Future<ApiResponse<Login200Response>> getCurrentUser(RoutingContext rc) {
    rc.user().attributes();
    var user = new Login200Response(new User("email", rc.request().getHeader("Authorization"), "user", "bio", "image"));
    return Future.succeededFuture(new ApiResponse<>(user));
  }

  public Future<ApiResponse<Login200Response>> login(Login body) {
    return Future.failedFuture(new HttpException(501));
  }

  public Future<ApiResponse<Login200Response>> updateCurrentUser(UpdateUser body) {
    return Future.failedFuture(new HttpException(501));
  }

}

File: src/main/kotlin/io/realworld/api/UserAndAuthenticationApiImpl.kt
package io.realworld.api

import io.realworld.extensions.mapTo
import io.realworld.model.request.CreateUserRequest
import io.realworld.model.request.LoginRequest
import io.realworld.model.request.UpdateCurrentUserRequest
import io.realworld.model.response.ApiResponse
import io.realworld.model.response.Login200Response
import io.realworld.service.UserService
import io.vertx.core.Future
import io.vertx.core.json.JsonObject

class UserAndAuthenticationApiImpl(
  private val userService: UserService
) : UserAndAuthenticationApi {
  override fun createUser(body: CreateUserRequest): Future<ApiResponse<Login200Response>> {
    return userService.createUser(JsonObject.mapFrom(body.user))
      .map { ApiResponse(data = it.mapTo()) }
  }

  override fun getCurrentUser(uid: Long): Future<ApiResponse<Login200Response>> {
    return userService.getCurrentUser(uid)
      .map { ApiResponse(data = it.mapTo()) }
  }

  override fun login(body: LoginRequest): Future<ApiResponse<Login200Response>> {
    return userService.login(body.user.email, body.user.password)
      .map<ApiResponse<Login200Response>?> { ApiResponse(data = it.mapTo()) }
      .onFailure {
        println(it.localizedMessage)
      }
  }

  override fun updateCurrentUser(uid: Long, body: UpdateCurrentUserRequest): Future<ApiResponse<Login200Response>> {
    return userService.updateUser(uid, JsonObject.mapFrom(body.user))
      .map { ApiResponse(data = it.mapTo()) }
  }
}
